spirv_reflect.hpp

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2018 Bradley Austin Davis
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SPIRV_CROSS_REFLECT_HPP
+#define SPIRV_CROSS_REFLECT_HPP
+
+#include "spirv_cross.hpp"
+#include <utility>
+#include <vector>
+
+namespace simple_json
+{
+class Stream;
+}
+
+namespace spirv_cross
+{
+class CompilerReflection : public Compiler
+{
+	using Parent = Compiler;
+
+public:
+	CompilerReflection(std::vector<uint32_t> spirv_)
+	    : Parent(move(spirv_))
+	{
+	}
+
+	CompilerReflection(const uint32_t *ir, size_t word_count)
+	    : Parent(ir, word_count)
+	{
+	}
+
+	void set_format(const std::string &format);
+	std::string compile() override;
+
+private:
+	static const char *execution_model_to_str(spv::ExecutionModel model);
+
+	void emit_entry_points();
+	void emit_types();
+	void emit_resources();
+	void emit_extensions();
+
+	void emit_type(const SPIRType &type, bool &emitted_open_tag);
+	void emit_type_member(const SPIRType &type, uint32_t index);
+	void emit_type_member_qualifiers(const SPIRType &type, uint32_t index);
+	void emit_type_array(const SPIRType &type);
+	void emit_resources(const char *tag, const std::vector<Resource> &resources);
+
+	std::string to_member_name(const SPIRType &type, uint32_t index) const;
+
+	std::shared_ptr<simple_json::Stream> json_stream;
```

Comment:
shared_ptr? Why not unique_ptr?

Reply:
You'll need to declare constructor and destructor as non-default/inline, then it should work.

Reply:
The code that fails to compile is in main.cpp (so no change to the `Stream` implementation in `spriv_reflect.cpp` is going to help), because that's where the `CompilerReflection` class is instantiated and destroyed.  Because it doesn't have visibility on the full `simple_json::Stream` type, it can't properly generate the destructor.  Explicitly adding a `~CompilerReflection` doesn't seem to help either.  

I can either move the `simple_json::Stream` full declaration into the spirv_reflect.hpp file (or into a new .cpp/.hpp pair in spirv-cross-util), making the full class declaration visible to main.cpp, or I can leave it as a `shared_ptr`.  Let me know.

Reply:
Ok, if that doesn't work, just keep shared_ptr.